File: Examples/test-suite/javascript/argcargvtest_runme.js
var test = require("argcargvtest");

const largs = ["hi", "hola", "hello"];
if (test.mainc(largs) != 3)
   throw "calling mainc failed";

const targs = ["hi", "hola"];
if (test.mainv(targs, 0) != "hi")
   throw "calling mainv failed";
if (test.mainv(targs, 1) != "hola")
   throw "calling mainv failed";
if (test.mainv(targs, 2) != "<<NULL>>")
   throw "calling mainv failed";

caughtException = false;
try {
   test.mainv("hello", 1);
} catch (err) {
   caughtException = true;
}
if (!caughtException) {
   throw "mainv without array should fail"
}

test.initializeApp(largs);

// Check that an empty array works.
const empty_args = [];
if (test.mainc(empty_args) != 0)
  throw "bad main typemap";
if (test.mainv(empty_args, 0) != "<<NULL>>")
   throw "calling mainv failed";

// Check that empty strings are handled.
const empty_string = ["hello", "", "world"];
if (test.mainc(empty_string) != 3)
  throw "bad main typemap";
if (test.mainv(empty_string, 0) != "hello")
  throw "bad main typemap";
if (test.mainv(empty_string, 1) != "")
  throw "bad main typemap";
if (test.mainv(empty_string, 2) != "world")
  throw "bad main typemap";
if (test.mainv(empty_string, 3) != "<<NULL>>")
  throw "bad main typemap";

File: Examples/test-suite/r/argcargvtest_runme.R
clargs <- commandArgs(trailing=TRUE)
source(file.path(clargs[1], "unittest.R"))
dyn.load(paste("argcargvtest", .Platform$dynlib.ext, sep=""))
source("argcargvtest.R")
cacheMetaData(1)

largs = c("hi", "hola", "hello")
unittest(3, mainc(largs))

targs = c("hi", "hola")
unittest("hi", mainv(targs, 0))
unittest("hola", mainv(targs, 1))
unittest("<<NULL>>", mainv(targs, 2))

# R convert the string to a string vector with a single string.
# So instead of exception we simply get null
unittest(NULL, mainv("hello", 1))

tryCatch({
    mainv(c(1, 2, 3), 1)
    # force an error if the previous line doesn't raise an exception
    stop("Test Failure A")
}, error = function(e) {
    if (e$message == "Test Failure A") {
      # Raise the error again to cause a failed test
      stop(e)
    }
    message("PASS")
})

initializeApp(largs)
initializeApp(largs, TRUE)

# Check that an empty array works.
empty_args = c()
unittest(0, mainc(empty_args))
unittest("<<NULL>>", mainv(empty_args, 0))

# check dispatcher with empty array.
initializeApp(empty_args)
initializeApp(empty_args, FALSE, TRUE)

# Check that empty strings are handled.
empty_string = c("hello", "", "world")
unittest(3, mainc(empty_string))
unittest("hello", mainv(empty_string, 0))
unittest("", mainv(empty_string, 1))
unittest("world", mainv(empty_string, 2))
unittest("<<NULL>>", mainv(empty_string, 3))
